src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java

```diff
@@ -0,0 +1,145 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class AssignmentOperationExpr implements OperationExpr {
+  public abstract VariableExpr variableExpr();
+
+  public abstract Expr valueExpr();
+
+  public abstract OperatorKind operatorKind();
+
+  @Override
+  public TypeNode type() {
+    return variableExpr().type();
+  }
+
+  @Override
+  public void accept(AstNodeVisitor visitor) {
+    visitor.visit(this);
+  }
+
+  public static AssignmentOperationExpr xorAssignmentWithExprs(
+      VariableExpr variableExpr, Expr valueExpr) {
+    return builder()
+        .setVariableExpr(variableExpr)
+        .setValueExpr(valueExpr)
+        .setOperatorKind(OperatorKind.ASSIGNMENT_XOR)
+        .build();
+  }
+
+  public static AssignmentOperationExpr multiplyAssignmentWithExprs(
+      VariableExpr variableExpr, Expr valueExpr) {
+    return builder()
+        .setVariableExpr(variableExpr)
+        .setValueExpr(valueExpr)
+        .setOperatorKind(OperatorKind.ASSIGNMENT_MULTIPLY)
+        .build();
+  }
+
+  private static Builder builder() {
+    return new AutoValue_AssignmentOperationExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    // Private setter.
+    abstract Builder setVariableExpr(VariableExpr variableExpr);
+
+    // Private setter.
+    abstract Builder setValueExpr(Expr valueExpr);
+
+    // Private setter.
+    abstract Builder setOperatorKind(OperatorKind operator);
+
+    abstract AssignmentOperationExpr autoBuild();
+
+    private AssignmentOperationExpr build() {
+      AssignmentOperationExpr assignmentOperationExpr = autoBuild();
+      TypeNode lhsType = assignmentOperationExpr.variableExpr().variable().type();
+      TypeNode rhsType =
+          assignmentOperationExpr.valueExpr() instanceof VariableExpr
+              ? ((VariableExpr) assignmentOperationExpr.valueExpr()).variable().type()
+              : assignmentOperationExpr.valueExpr().type();
+      OperatorKind operator = assignmentOperationExpr.operatorKind();
+
+      // Check if the variable exprs have been declared, if yes, throw error.
+      Preconditions.checkState(
+          !assignmentOperationExpr.variableExpr().isDecl()
+              && (assignmentOperationExpr.valueExpr() instanceof VariableExpr
+                  ? !((VariableExpr) assignmentOperationExpr.valueExpr()).isDecl()
+                  : true),
+          String.format(
+              "Variable `%s` should not be declaration in the variable expression.",
+              assignmentOperationExpr.variableExpr().variable().name()));
+
+      // TYPE_CHECK_ERROR_MSG is type checking error message for operators.
+      final String TYPE_CHECK_ERROR_MSG =
```

Comment:
ALL_CAPS variable names are usually reserved for `private static final` variables, otherwise we use lowerCamelCase.

Reply:
Good to know. Updated.